File: src/main/java/com/xebia/xboson/ParsedDocument.java
package com.xebia.xboson;

import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by xviuda on 22-01-16.
 */
public class ParsedDocument {
    private final Document doc;

    public ParsedDocument(Document doc) {
        this.doc = doc;
    }


    public List<ParsedElement> getElements() {
        Elements elements = doc.getElementsByTag("input");


        // Apply rules to make elements unique

        return elements.stream().map(e -> new ParsedElement(identifier(e), selector(e))).collect(Collectors.toList());
    }

    private String selector(Element e) {
        return null;
    }

    private String identifier(Element e) {
        return null;
    }
}

File: src/main/kotlin/com/xebia/xboson/ParsedDocument.kt
package com.xebia.xboson

import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import org.jsoup.select.Elements

import java.util.ArrayList
import java.util.Optional
import java.util.stream.Collectors

class ParsedDocument(private val doc: Document) {
    private val selectionRules = listOf(
        NameSelectionRule(),
        TagSelectionRule(),
        IdSelectionRule()
    )

    // Apply rules to make elements unique
    val elements: List<ParsedElement>
        get() {
            val elements = doc.select("input, button, a, textarea")

//            return elements.map { e -> match(e) } .filterNotNull();
            return elements.asSequence().map { e -> match(e) } .filterNotNull().toList();
        }

    private fun match(e: Element): ParsedElement? {
        return selectionRules.asSequence().map { rule -> rule.match(doc, e) } .filterNotNull().firstOrNull()
//        for (rule in selectionRules) {
//            val m = rule.match(doc, e)
//            if (m != null) {
//                return m
//            }
//        }
//        return null
    }

}
